src/lib.rs

```diff
@@ -1,19 +1,521 @@
+}
+
+#[derive(Debug, Clone)]
+#[must_use]
+pub struct Verifier {
+    message: Message,
+    valid_at: Time,
+
+    statement: Option<String>,
+    statement_mode: Verify,
+
+    expiration_time_mode: Verify,
+
+    not_before_mode: Verify,
+
+    domain: Authority,
+    uri: UriString,
+    chain_id: u64,
+}
+
+impl Verifier {
+    pub fn nonce(&self) -> &str {
+        &self.message.nonce
+    }
+
+    /// Specify the time to use when validating the message. Defaults to the current time.
+    pub fn valid_at(mut self, now: OffsetDateTime) -> Self {
+        self.valid_at = Time::Exact(now);
+        self
+    }
+
+    /// Choose when to verify the statement. Defaults to [`Verify::Always`].
+    pub fn statement(mut self, verify: Verify) -> Self {
+        self.statement_mode = verify;
+        self
+    }
+
+    /// Choose when to verify the expiration time. Defaults to [`Verify::WhenPresent`].
+    pub fn expiration_time(mut self, verify: Verify) -> Self {
+        self.expiration_time_mode = verify;
+        self
+    }
+
+    /// Choose when to verify the not before time. Defaults to [`Verify::WhenPresent`].
+    pub fn not_before(mut self, verify: Verify) -> Self {
+        self.not_before_mode = verify;
+        self
+    }
+
+    pub fn verify(self, nonce: &str, signature: &[u8; 65]) -> Result<Message, VerificationError> {
+        self.message.verify_eip191(&signature)?;
+
+        let valid_at = self.valid_at.resolve().expect("`valid_at` out of range");
+
+        if self.domain != self.message.domain {
+            return Err(VerificationError::DomainMismatch);
+        }
+
+        if self.uri != self.message.uri {
+            return Err(VerificationError::UriMismatch);
+        }
+
+        if self.chain_id != self.message.chain_id {
+            return Err(VerificationError::ChainIdMismatch);
+        }
+
+        if nonce != self.message.nonce {
+            return Err(VerificationError::NonceMismatch);
+        }
+
+        // Have to verify the statement because it could contain, for example, Terms of Service.
+        let cmp_stmt = |expect, actual| expect == actual;
+        if !self
+            .statement_mode
+            .verify(self.statement, self.message.statement.clone(), cmp_stmt)
+        {
+            return Err(VerificationError::StatementMismatch);
+        }
+
+        let cmp_exp = |valid_at, expires_at| expires_at >= valid_at;
+        if !self.expiration_time_mode.verify(
+            Some(valid_at),
+            self.message.expiration_time.clone(),
+            cmp_exp,
+        ) {
+            return Err(VerificationError::Time);
+        }
+
+        let cmp_nbf = |valid_at, not_before| not_before < valid_at;
+        if !self
+            .not_before_mode
+            .verify(Some(valid_at), self.message.not_before.clone(), cmp_nbf)
+        {
+            return Err(VerificationError::Time);
+        }
+
+        Ok(self.message)
+    }
+}
+
+#[derive(Debug, Clone)]
+#[must_use]
+pub struct Builder {
```

Comment:
We do like the idea of a builder, but we would prefer one that makes ignoring the optional fields more ergonomic whilst keeping compile time guarantees.  Maybe `typed_builder` is what we need?

Reply:
I _love_ `typed_builder`!

Wasn't sure you'd be amenable to adding the dependency though. If you are, I'll most certainly use it.

Reply:
IMO I think we are fine to add `typed_builder`, but it should be behind a feature flag.